rust/flatbuffers/src/builder.rs

```diff
@@ -235,6 +240,21 @@ impl<'fbb> FlatBufferBuilder<'fbb> {
         WIPOffset::new(o.value())
     }
 
+    #[inline]
+    pub fn create_shared_string<'a: 'b, 'b>(&'a mut self, s: &'b str) -> WIPOffset<&'fbb str> {
+        self.assert_not_nested(
+            "create_shared_string can not be called when a table or vector is under construction",
+        );
+        match self.strings_map.get(s) {
+            Some(offset) => return WIPOffset::new(*offset),
+            None => {
+                let offset = self.create_byte_string(s.as_bytes()).value();
+                self.strings_map.insert(String::from(s), offset);
+                return WIPOffset::new(offset)
+            }
```

Comment:
Makes a copy of str to String since that's the only workaround I found. i am not sure if there is another way to do it

Reply:
I bet the reason `str` didn't work was because it takes an immutable reference to builder's `owned_buf` but while building the flatbuffer, we need to modify `owned_buf`, potentially invalidating the reference due to vector reallocation. To avoid the copy, you could maintain a vector of WIPOffset that's sorted by the string. This is the approach I used in flexbuffers.

https://github.com/google/flatbuffers/blob/4e79d129cb9018fdf6584cc1fef6a8d2db41715e/rust/flexbuffers/src/builder/mod.rs#L170

Reply:
So from what I understood from that code, is that we would be comparing it with the string saved in the buffer directly instead of a `HashMap`. where we take each WIPOffset in the vector and go looking for a similar string in the buffer according to the address given in the WIPOffset. But what would be more expensive when it comes to large amount of data, a copy, or iterating over a vector?

Reply:
Yep, you understand correctly. Iterating over the vector isn't expensive since its sorted. The expensive part is inserting a new string. I guess this depends on the length of strings versus number of unique strings in the workload. Some profiling required I guess.

Note that Rust by default uses a crypographic but slow hash and the C++ flexbuffers implementation uses `std::set`, which I think is btree based. When I decided to go with a sorted vector in flexbuffers instead of a hashset, I used a workload that had like ~100 unique keys each ~25 char long and assumed that was typical. I'm not sure when the hashset/btreeset gets faster than the sorted vector.

I think there should be a way, perhaps with unsafe, to get good asymptotics without copying strings but that'll be a bit of work.